src/mlpack/methods/ann/layer/memory_head_impl.hpp

```diff
@@ -0,0 +1,482 @@
+
+  // kT non linearity.
+  kTNonLinear = new TanHLayer<>();
+
+  network.push_back(inputLinear);
+  network.push_back(kTNonLinear);
+
+  allZeros = arma::zeros<arma::mat>(outSize, 1);
+
+  prevWeights.push_back(std::move(arma::mat(allZeros.memptr(),
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  weightsBackwardIterator = prevWeights.end();
+
+  prevError = arma::zeros<arma::mat>((memSize) + (1) + (1) +
+      (2 * shiftSize + 1) + (1), 1);
+
+  bWdash = lWDash.end();
+  bGammaT = lGammaT.end();
+  bWTilde = lWTilde.end();
+  bShiftMatrix = lShiftMatrix.end();
+  bWg = lWg.end();
+  bSt = lSt.end();
+  bGt = lGt.end();
+  bWe = lWe.end();
+  bWc = lWc.end();
+  bBt = lBt.end();
+  bCosineT = lConsineT.end();
+  bMemoryHistory = memoryHistory.end();
+}
+
+template <typename InputDataType, typename OutputDataType>
+MemoryHead<InputDataType, OutputDataType>::~MemoryHead()
+{
+  boost::apply_visitor(deleteVisitor, inputLinear);
+  boost::apply_visitor(deleteVisitor, kTNonLinear);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void MemoryHead<InputDataType, OutputDataType>::Forward(arma::Mat<eT>&& input,
+                                                        arma::Mat<eT>&& output)
+{
+  const arma::mat& memory = memoryHistory.back();
+
+  // Pass the input through linear layer.
+  boost::apply_visitor(ForwardVisitor(std::move(input), std::move(
+      boost::apply_visitor(outputParameterVisitor, inputLinear))),
+      inputLinear);
+
+  arma::mat& lOutput = boost::apply_visitor(outputParameterVisitor,
+    inputLinear);
+
+  // Build kT with non linearity.
+  boost::apply_visitor(ForwardVisitor(std::move(lOutput.submat(0, 0,
+      memSize - 1, 0)), std::move(boost::apply_visitor(outputParameterVisitor,
+      kTNonLinear))), kTNonLinear);
+  const arma::mat& kT = boost::apply_visitor(outputParameterVisitor,
+      kTNonLinear);
+
+  // Build bT with non linearity
+  lBt.push_back(bTNonLinear.Fn(lOutput(memSize, 0)));
+  const double& bT = lBt.back();
+
+  // Build gT with non linearity
+  lGt.push_back(gTNonLinear.Fn(lOutput(memSize + 1, 0)));
+  const double& gT = lGt.back();
+
+  // Build sT with non linearity
+  arma::vec temp = arma::exp(lOutput.submat(memSize + 2, 0,
+    memSize + 2 + 2 * shiftSize, 0));
+  temp = temp / arma::as_scalar(arma::sum(temp));
+  lSt.push_back(std::move(temp));
+  const arma::vec& sT = lSt.back();
+
+  // Build gammaT with non linearity
+  lGammaT.push_back(gammaTNonLinear.Fn(arma::as_scalar(
+    lOutput.submat(memSize + 2 + 2 * shiftSize + 1, 0,
+    memSize + 2 + 2 * shiftSize + 1, 0))));
+  const double& gammaT = lGammaT.back();
+
+  // Perform cosine similarity with memory content
+  lConsineT.push_back(arma::normalise(memory, 2, 1) * arma::normalise(kT));
+  const arma::vec& cosSimilarity = lConsineT.back();
+
+  // Build wC with bT and softmax
+  lWe.push_back(arma::exp(bT * cosSimilarity));
+  const arma::vec& wE = lWe.back();
+
```

Comment:
We should remove the line here, since ``lwe`` doesn't contain the softmax, ``lWc`` does.